### ISSUE TITLE
Private fields on resources

### ISSUE THREAD

**Issue by leplatrem** [mozilla-services/cliquet#547]
Source: https://github.com/mozilla-services/cliquet/issues/543#issuecomment-153396221

It could be useful for some use-cases to have a schema option to specify private fields (the same way we specify readonly fields)

```python
class BookmarkSchema(resource.ResourceSchema):
    url = colander.SchemaNode(colander.String(), validator=colander.url)
    title = colander.SchemaNode(colander.String())
    comment = colander.SchemaNode(colander.String())

    class Options:
        private_fields = ('comment',)
```

When specified, these fields wouldn't appear when the collection/record is requested anonymously.

Maybe @MrChoclate could help specifying this and contribute it :)

**Comment by MrChoclate** [comment]
Well private_fields have a meaning if the resource is shared with someone else. Then I see ways to implement it:
- Only the author can see them (How do you known the author?).
- Only someone with write permissions can see them (write to read a field.. not that good but why not?).
- Specify a new permission: not good at all.
- They are hided on collections but we show them to the one who got permission to read the record (my favorite one, but this is my use case).

**Comment by leplatrem** [comment]
Thanks for the follow-up !

> They are hidden on collections but we show them to the one who got permission to read the record

I'm not against this one. Maybe *private* is not the right term... *restricted*, *hidden*,... can't find a better one though :)

**Comment by MrChoclate** [comment]
I got a new but similar use case:

We have a Lesson resource:

```python
class LessonSchema(resource.ResourceSchema):
    title = colander.SchemaNode(colander.String())
    big_content = colander.SchemaNode(colander.String())
```

The content of a lesson is very big, then when we got the collection of lessons we do not want to download the big_content of all lessons, we only want to download the title, then if we are interested in a lesson, we look up for the record and download the big_content.

I think this is important if our user has a poor Internet connection.

We could have a record_only_fields propriety that hide this big fields on collection. What do you think?

I would do a PR when I got time for the hidden_fields: fields that are hidden to the one that do not have permission to read the record.

**Comment by leplatrem** [comment]
> We could have a record_only_fields propriety that hide this big fields on collection. What do you think?

I think that what you describe here in this second use-case is the *partial response* feature, described in #384. We could, as you suggest, add an option on schema/resource to force which fields are served, but I think that I prefer the approach where the client specifies them like `GET /lessons?_fields=title`.

As for the hidden_fields contribution, please do! Do not hesitate to publish your work in progress and gather feedback along the way! Thanks a lot for your involvement!

**Comment by MrChoclate** [comment]
Well, what I did feels like a dead-end.

- I added hidden_fields to `RessourceSchema.Options`.
- I hide this fields on `MemoryBasedStorage.extract_record_set()` (then it is hidden on collection)
- Added all the hidden_fields args from `UserRessource` to `extract_record_set()` so it works.

I got `cliquet.champion_read_principals = system.Everyone`. So everyone can see all champions except the hidden field. This is great. But everyone can see the every record so it is pointless.

Permission check is done before storage calls.
There is no permission check in the storage. Adding a permission check in the storage.get feels wrong.
Is my design bad? Where should I do the permission check and hide my fields?

I did not do a PR but my code is available here: 
https://github.com/MrChoclate/cliquet/tree/feat/hidden-fields

**Comment by leplatrem** [comment]
I overlooked your branch, but couldn't see the main logic you tried to implement.

So what about: 

- First start by implementing #384, it will give us the base for fields filtering in the resource code
- Then, when `hidden_fields` option is set on model, force the partial content for the collection using the code implemented for #384 
- Then, do not force the partial content when the current user has the `write` permission (for example)

The last step is indeed the one that deserves integrating the permission backend (or the authorization context). But we can think of something clean meanwhile :) Like for example, adding the user permissions on the record in the authorization context, and then in the resource check that `write` is in this `self.context.record_permissions` before choose to filter or not...

Don't hesitate to come back to us if some parts are still unclear!

Thanks again ! Enjoy !

**Comment by tarekziade** [comment]
@MrChoclate  we're moving everything to Kinto - if you are still active on this topic please reopen a bug there. Thanks a lot !
